WPThemeReview/Sniffs/CoreFunctionality/PrefixAllGlobalsSniff.php

```diff
@@ -0,0 +1,93 @@
+ *
+ * @package WPTRT\WPThemeReview
+ * @link    https://github.com/WPTRT/WPThemeReview
+ * @license https://opensource.org/licenses/MIT MIT
+ */
+
+namespace WPThemeReview\Sniffs\CoreFunctionality;
+
+use WordPressCS\WordPress\Sniffs\NamingConventions\PrefixAllGlobalsSniff as WPCSPrefixAllGlobalsSniff;
+use WordPressCS\WordPress\Sniffs\Files\FileNameSniff;
+
+/**
+ * Verify that everything defined in the global namespace is prefixed with a theme specific prefix.
+ *
+ * This sniff extends the upstream WPCS PrefixAllGlobalsSniff. The differences are:
+ * - For non-prefixed global variables, an error will only be thrown when the variable is created outside of a theme template file.
+ *
+ * @link  https://github.com/WPTRT/WPThemeReview/issues/205
+ * @link  https://github.com/WPTRT/WPThemeReview/issues/201
+ * @link  https://github.com/WPTRT/WPThemeReview/issues/200
+ *
+ * @since 0.2.0
+ */
+class PrefixAllGlobalsSniff extends WPCSPrefixAllGlobalsSniff {
+
+	/**
+	 * The list of allowed folders to check the file path against.
+	 *
+	 * This array can be extended in the custom ruleset.
+	 * If you are using PHPCS prior to the version 3.4.0 you would overwrite the current value by setting:
+	 *
+	 * <property name="allowed_folders" type="array" values="template-parts,templates,partials,my-template-folder">
+	 *
+	 * If you are using PHPCS 3.4.0 or greater you can extend this array by setting extends="true" attribute:
+	 *
+	 * <property name="allowed_folders" type="array" extends="true" values="my-template-folder">
+	 *
+	 * @var array
+	 */
+	public $allowed_folders = [
+		'template-parts',
+		'templates',
+		'partials',
+	];
+
+	/**
+	 * Check that defined global variables are prefixed.
+	 *
+	 * This overloads the parent method to allow for non-prefixed variables to be declared
+	 * in template files as those are included from within a function and therefore would be
+	 * local to that function.
+	 *
+	 * @since 0.2.0
+	 *
+	 * @param int $stackPtr The position of the current token in the stack.
+	 *
+	 * @return int|void Integer stack pointer to skip forward or void to continue
+	 *                  normal file processing.
+	 */
+	protected function process_variable_assignment( $stackPtr ) {
+
+		// Usage of `strip_quotes` is to ensure `stdin_path` passed by IDEs does not include quotes.
+		$file = $this->strip_quotes( $this->phpcsFile->getFileName() );
+
+		$fileName = basename( $file );
+
+		// Don't process in case of a template file or folder.
+		if ( preg_match( FileNameSniff::THEME_EXCEPTIONS_REGEX, $fileName ) === 1 && $this->is_from_allowed_folder( $file ) ) {
+			return;
+		}
+
+		return parent::process_variable_assignment( $stackPtr );
+	}
+
+	/**
+	 * Checks if the given file path is located in the $allowed_folders array
+	 *
+	 * @param  string $path Full path of the sniffed file.
+	 * @return boolean
+	 */
+	private function is_from_allowed_folder( $path ) {
+		foreach ( $this->allowed_folders as $folder ) {
```

Comment:
If the default for `$allowed_folders` is going to be an empty array, I'd suggest adding the below code at the top of the function.
```php
if ( empty( $this->allowed_folders ) || ! is_array( $this->allowed_folders ) ) {
    return false

}
```